test/stdlib/OSLogPrototypeExecTest.swift

```diff
@@ -66,4 +66,361 @@ if #available(OSX 10.12, iOS 10.0, watchOS 3.0, tvOS 10.0, *) {
+internal let bitsPerByte = 8
+
+/// A struct that exposes methods for checking whether a given byte buffer
+/// conforms to the format expected by the os_log ABI. This struct acts as
+/// a specification of the byte buffer format.
+internal struct OSLogBufferChecker {
+
+  internal let buffer: UnsafeBufferPointer<UInt8>
+
+  internal init(_ byteBuffer: UnsafeBufferPointer<UInt8>) {
+    buffer = byteBuffer
+  }
+
+  /// Bit mask for setting bits in the peamble. The bits denoted by the bit
+  /// mask indicate whether there is an argument that is private, and whether
+  /// there is an argument that is non-scalar: String, NSObject or Pointer.
+  internal enum PreambleBitMask: UInt8 {
+    case privateBitMask = 0x1
+    case nonScalarBitMask = 0x2
+  }
+
+  /// Check the summary bytes of the byte buffer.
+  /// - Parameters:
+  ///  - argumentCount: number of arguments expected to be in the buffer.
+  ///  - hasPrivate: true iff there exists a private argument
+  ///  - hasNonScalar: true iff there exists a non-scalar argument
+  internal func checkSummaryBytes(
+    argumentCount: UInt8,
+    hasPrivate: Bool,
+    hasNonScalar: Bool
+  ) {
+    let preamble = buffer[0]
+    let privateBit = preamble & PreambleBitMask.privateBitMask.rawValue
+    expectEqual(hasPrivate, privateBit != 0)
+
+    let nonScalarBit = preamble & PreambleBitMask.nonScalarBitMask.rawValue
+    expectEqual(hasNonScalar, nonScalarBit != 0)
+
+    expectEqual(argumentCount, buffer[1])
+  }
+
+  /// The possible values for the argument flag as defined by the os_log ABI.
+  /// This occupies four least significant bits of the first byte of the
+  /// argument header. Two least significant bits are used to indicate privacy
+  /// and the other two bits are reserved.
+  internal enum ArgumentFlag: UInt8 {
+    case privateFlag = 0x1
+    case publicFlag = 0x2
+  }
+
+  /// The possible values for the argument type as defined by the os_log ABI.
+  /// This occupies four most significant bits of the first byte of the
+  /// argument header.
+  internal enum ArgumentType: UInt8 {
+    case scalar = 0, count, string, pointer, object
+    // TODO: include wide string and errno here if needed.
+  }
+
+  /// Check the encoding of an argument in the byte buffer starting from the
+  /// `startIndex`.
+  ///  - precondition: `T` must be a type that is accepted by os_log ABI.
+  private func checkArgument<T>(
+    startIndex: Int,
+    size: UInt8,
+    flag: ArgumentFlag,
+    type: ArgumentType,
+    expectedData: T
+  ) {
+    let argumentHeader = buffer[startIndex]
+    expectEqual((type.rawValue << 4) | flag.rawValue, argumentHeader)
+
+    expectEqual(size, buffer[startIndex + 1])
+
+    // Check every byte of the payload.
+    withUnsafeBytes(of: expectedData) { expectedBytes in
+      for i in 0..<Int(size) {
+        // Argument data starts after the two header bytes.
+        expectEqual(expectedBytes[i], buffer[startIndex + 2 + i])
```

Comment:
Probably to make a failure more useful, add a message to the `expectEqual` call, containing the index at which the mismatch happened.